src/components/select/select.js

```diff
@@ -893,10 +893,37 @@ function SelectMenuDirective($parse, $mdUtil, $mdConstant, $mdTheming) {
       var newVal = self.isMultiple ? values : values[0];
       var prevVal = self.ngModel.$modelValue;
 
-      if (usingTrackBy ? !angular.equals(prevVal, newVal) : (prevVal + '') !== newVal) {
+      if (usingTrackBy ? !equals(prevVal, newVal) : (prevVal + '') !== newVal) {
```

Comment:
Because the new `equals` function relies on `self.hashGetter`, the `self.hashGetter` method needs to be redefined (potentially), depending on if `usingTrackBy` has changed value.


For instance, if on initialization, `$mdUtil.getModelOption(self.ngModel, 'trackBy')` evaluates to `false` then later when running `self.refreshViewValue`, `$mdUtil.getModelOption(self.ngModel, 'trackBy')` evaluates to `true`, the `self.hashGetter` function that is called is not the one that is expected.


This was not an issue in the past, as the `angular.equals` does not rely on the `hashGetter` method that is defined.  This issue could also come to light in any other usage of `self.hasGetter` throughout the object.

Reply:
Ooops! If changing `trackBy` after initialization is supported, it will get a little messy :grin:
The last `usingTrackBy` value should be kept track of and checked everywhere before using `self.hashGetter()` (afaict).

Reply:
I guess that we also need to add a test for changing `trackBy` on `md-select`. I never imagined a use case where you would want to change `trackBy` over time.

Reply:
Something like this would make sense?

```js
self.hashGetter = function(value) {
  var trackByOption = $mdUtil.getModelOption(ngModel, 'trackBy');

  if (trackByOption) {
    return $parse(trackByOption)($scope, { $value: value });

    // If the user doesn't provide a trackBy, we automatically generate an id for every
    // value passed in
  } else if (angular.isObject(value)) {
      return 'object_' + (value.$$mdSelectId || (value.$$mdSelectId = ++selectNextId));
  }
  return value;
};
```
Basically I evaluate `ngModelOption.trackBy` each time.

Clearly it's not so performant as it was.

Reply:
I was trying to figure out if there is a way to watch/observe changes to `ng-model-options` and I found this:

From the AngularJS docs:
> The ngModelOptions expression is only evaluated once when the directive is linked; it is not watched for changes. However, it is possible to override the options on a single ngModel.NgModelController instance with [NgModelController#$overrideModelOptions](https://docs.angularjs.org/api/ng/type/ngModel.NgModelController#$overrideModelOptions).

The linked docs specify
> Note: this function only affects the options set on the `ngModelController`, and not the options on the `ngModelOptions` directive from which they might have been obtained initially.

That should be OK if we're watching the controller (as we are now) but it means that watching the `ng-model-options` attribute directive for changes would be a bad idea.

@josephperrott can you confirm that the failures are being caused by calls to `NgModelController#$overrideModelOptions` in that app?

@gkalpak do you see any more performant alternatives than the suggestion above?

Reply:
No, I don't see any great alternative. I would still check `trackBy` outside of `hashGetter`, because that gets called multiple times inside certain functions. I would check it at the top of functions that call `hashGetter()`.

Note that AngularJS generally does not check for changes in `modelOptions`, but `trackBy` is an `ngMaterial` specific option (and there is nothing stopping `ngMaterial` from responding to `modelOptions` changes anyway.

BTW, I doubt `$overrideModelOptions()` is relevant here.

Reply:
The failures I have found do something to this effect:

```html
<md-select ng-model="modelName ng-model-options="{
    trackBy: 'ctrl.hasTracker() ? ctrl.tracker() : undefined'}">
  <md-option ng-value="something" ng-repeat="something in ctrl.somethings">
    {{something}}
  </md-option>
</md-select>
```

As @gkalpak noted, `$overrideModelOptions()` isn't really in play here.

Reply:
@free-easy OK, let's try your approach in https://github.com/angular/material/pull/11156#discussion_r177416656 and see if it gets through presubmit testing. I agree that things get a bit too messy when having to scatter calls around different functions.

If it becomes a performance problem that users notice, we'll drop support for changing `trackBy` after initialization in 1.2.0.

Reply:
@Splaktar It seems that it doesn't work anyway since angular.js doesn't update ngModelOptions.

```js
ng-model-options="{ trackBy: 'ctrl.hasTracker() ? ctrl.tracker() : undefined' }"
```
The ternary expression seems to be evaluated only once.

Reply:
@free-easy are you saying that `$mdUtil.getModelOption(self.ngModel, 'trackBy')` mentioned in https://github.com/angular/material/pull/11156#discussion_r177183110 can't currently change after initialization as described there?

Reply:
Exactly

Reply:
Exactly